felix/nftables/table.go

```diff
@@ -0,0 +1,1173 @@
+		allHashes[i+offset] = hash
+	}
+
+	offset = len(insertedRules) + numNonCalicoRules
+	for i, hash := range ourAppendedHashes {
+		allHashes[i+offset] = hash
+	}
+	return
+}
+
+// getHashesAndRulesFromDataplane loads the current state of our table. It parses out the hashes that we
+// add to rules and, for chains that we insert into, the full rules. The 'hashes' map contains an entry for each chain
+// in the table. Each entry is a slice containing the hashes for the rules in that table. Rules with no hashes are
+// represented by an empty string. The 'rules' map contains an entry for each non-Calico chain in the table that
+// contains inserts. It is used to generate deletes using the full rule, rather than deletes by line number, to avoid
+// race conditions on chains we don't fully control.
+func (t *nftablesTable) getHashesAndRulesFromDataplane() (hashes map[string][]string, rules map[string][]*knftables.Rule) {
+	retries := 3
+	retryDelay := 100 * time.Millisecond
+
+	// Retry a few times before we panic.  This deals with any transient errors and it prevents
+	// us from spamming a panic into the log when we're being gracefully shut down by a SIGTERM.
+	for {
+		t.onStillAlive()
+		hashes, rules, err := t.attemptToGetHashesAndRulesFromDataplane()
+		if err != nil {
+			countNumSaveErrors.Inc()
+			var stderr string
+			if ee, ok := err.(*exec.ExitError); ok {
+				stderr = string(ee.Stderr)
+			}
+			t.logCxt.WithError(err).WithField("stderr", stderr).Warn("nftables command failed")
+			if retries > 0 {
+				retries--
+				t.timeSleep(retryDelay)
+				retryDelay *= 2
+			} else {
+				t.logCxt.Panic("nftables command failed after retries")
+			}
+			continue
+		}
+
+		return hashes, rules
+	}
+}
+
+// attemptToGetHashesAndRulesFromDataplane reads nftables state and loads it into memory.
+func (t *nftablesTable) attemptToGetHashesAndRulesFromDataplane() (hashes map[string][]string, rules map[string][]*knftables.Rule, err error) {
+	startTime := t.timeNow()
+	defer func() {
+		saveDuration := t.timeNow().Sub(startTime)
+		t.peakNftablesReadTime = t.peakNftablesReadTime * 99 / 100
+		if saveDuration > t.peakNftablesReadTime {
+			t.logCxt.WithField("duration", saveDuration).Debug("Updating nftables peak duration.")
+			t.peakNftablesReadTime = saveDuration
+		}
+	}()
+
+	t.logCxt.Debug("Attmempting to get hashes and rules from nftables")
+
+	hashes = make(map[string][]string)
+	rules = make(map[string][]*knftables.Rule)
+
+	ctx, cancel := context.WithTimeout(context.Background(), t.contextTimeout)
+	defer cancel()
+
+	// Add chains. We need to query this separately, as chains may exist without rules.
+	allChains, err := t.nft.List(ctx, "chain")
+	if err != nil {
+		if isNotFound(err) {
+			err = nil
+			return
+		}
+		return nil, nil, err
+	}
+	for _, chain := range allChains {
+		hashes[chain] = []string{}
+		rules[chain] = []*knftables.Rule{}
+	}
+
+	// List rules and extract the hashes.
+	allRules, err := t.nft.ListRules(ctx, "")
```

Comment:
`""` only works by coincidence here; you should submit a knftables patch to document/test that (and update `Fake` to support it).

(I eventually want to totally redo the List APIs, but I'm not sure how yet, so it's probably not worth trying to make the API any nicer to support this case yet.)

Reply:
It's not _quite_ coincidence :grin: 

https://github.com/kubernetes-sigs/knftables/pull/7

Reply:
I missed the fake support in the first PR, though: https://github.com/kubernetes-sigs/knftables/pull/12

Reply:
doh, forgot about that PR and then checked my current checked-out copy of that code, which predates your PR, and I saw that it didn't look like it would work and wasn't documented as working. :slightly_smiling_face: